### ISSUE TITLE
keras.LSTM in Tensorflow built with mkl is 5 times slower on CPU

### ISSUE THREAD

**Issue by npuichigo** [tensorflow/tensorflow#33138]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 16.04
- TensorFlow installed from (source or binary): source (with or without mkl)
- TensorFlow version (use command below): 2.0.0
- Python version: 3.6.8
- Bazel version (if compiling from source): 0.20.0
- GCC/Compiler version (if compiling from source):  7.4.0
- CUDA/cuDNN version: 7.6.2

**Describe the current behavior**
keras.LSTM in Tensorflow built with mkl is 5 times slower on CPU

**Other info / logs**
tf.matmul and tf.keras.Conv2D are also tested and mkl version is faster.

**Comment by Leslie-Fang** [comment]
Hi @npuichigo 
Did you built TF on source code or use the pre-built wheel package?
Could you give more information about the tensorflow version and share the benchmark code?
I can help to re-produce the issue.

**Comment by npuichigo** [comment]
@Leslie-Fang I built the python wheel packages with or without mkl both from source code (tf 2.0.0). I simply compared the time used for tf.keras.LSTM(..., return_sequences=True) with or without mkl and find the difference. It's nice that you can help to re-produce the issue.

**Comment by Leslie-Fang** [comment]
Sure, I can try the code in latest master branch.
But I am not familiar with this tf.keras.LSTM(..., return_sequences=True) API.
Could you give me a snippet of code?

**Comment by npuichigo** [comment]
```python
import timeit

import numpy as np
import tensorflow as tf
from tensorflow import keras


if __name__ == "__main__":
    lstm = keras.layers.LSTM(1024, return_sequences=True)
    inputs = tf.random.normal((1, 100, 1024))
    time_used = np.mean(timeit.repeat(lambda: lstm(inputs), repeat=10, number=1))
    print("Time used: {}s".format(time_used))
```

**Comment by Leslie-Fang** [comment]
Hi @npuichigo 
Here is the result I got:
No-MKL: Time used: 0.8596639971015975s
MKLDNN: Time used: 0.8698679601977346s

**Comment by npuichigo** [comment]
@Leslie-Fang Did you enable all the SSE/AVX features when compiling the packages?
The result I got:
No-MKL: Time used: 0.2749028032645583s
MKLDNN: Time used: 1.2454360516741871s

**Comment by Leslie-Fang** [comment]
> @Leslie-Fang Did you enable all the SSE/AVX features when compiling the packages?
> The result I got:
> No-MKL: Time used: 0.2749028032645583s
> MKLDNN: Time used: 1.2454360516741871s

Good suggestion.  I can reproduce your performance data now.
I will do some profile and let you know the result.

**Comment by feihugis** [comment]
I also met the same problem.

**Comment by jvishnuvardhan** [comment]
@npuichigo I think this is similar to this [issue](https://github.com/tensorflow/tensorflow/issues/33146). Thanks!

**Comment by penpornk** [comment]
@npuichigo @Leslie-Fang That's because the code is run in eager mode. TensorFlow-MKL doesn't support many operations in eager yet. See the list of supported ops in eager mode [here](https://github.com/tensorflow/tensorflow/blob/8cb8627abb5ef83a6fba34f8fd0e4ee430562eb1/tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc#L79-L87). (There will be more soon in later releases.)

@TensorFlow-MKL Maybe we should make sure that the ops that can't be rewritten doesn't get slower than normal TF?

**Comment by feihugis** [comment]
@penpornk I met a similar performance issue on TF `v1.14.0` when running my LSTM-based model on [intel-tensorflow docker](https://hub.docker.com/r/intelaipg/intel-optimized-tensorflow/). The profiling file can be found [here](https://gist.github.com/feihugis/f20529b25fe17ba3b46d31580ba7ef92).

**Comment by Leslie-Fang** [comment]
@penpornk Thanks for your information.
https://github.com/tensorflow/tensorflow/blob/308a0d42c5bb7a37c259a73a282c5c0338aecb8d/tensorflow/python/keras/layers/recurrent_v2.py#L1262-L1279
Here is the block which LSTM is repeated calculated, MKL version do spend more time in the calculation of this code block.
From the log I see matmul has been rewrited to _mklmatmul.

**Comment by jojivk73** [comment]
@npuichigo, Thanks for all the information. Looking into this, I will update once I find the root cause.

**Comment by AStupidBear** [comment]
@Leslie-Fang @jojivk73 @penpornk Why dosen't tensorflow directly call MKL-DNN's LSTM primitives? That's the fastest approach.

**Comment by jojivk73** [comment]
The current performance is better with DNNL 1.x.
   MKL Time : 0.2231828489s

**Comment by AStupidBear** [comment]
@jojivk73 Which version of tensorflow can produce this result?

**Comment by jojivk73** [comment]
@AStupidBear The fix is in latest branch TF 2.x.

**Comment by penpornk** [comment]
@jojivk73 Would you mind pasting the commit hash here? I'd like to check if it made it into the TF 2.3 branch. Thank you!

**Comment by jojivk73** [comment]
@penpornk This is the related oneDNN [issue ](https://github.com/oneapi-src/oneDNN/issues/525#issuecomment-575375623) :  [commit](https://github.com/oneapi-src/oneDNN/commit/cf7edb6575b255a2fee54d99d0f2eef0e761939a)

**Comment by penpornk** [comment]
Thank you! TF-MKL 2.3 is currently using oneDNN v1.4 which doesn't have this fix. There's a plan to cherrypick a change to v1.5 if possible, but it seems v1.5 doesn't have this fix either. Do you plan to patch this into v1.5?

cc: @agramesh1

**Comment by agramesh1** [comment]
Hi @penpornk  The fix is in oneDNN 1.4 so we should be fine for  TF2.3 regarding this bug. The oneDNN commit is also in oneDNN 1.2 which was used in TF2.2.  So TF2.2 should resolve the issue.

**Comment by jojivk73** [comment]
Hi @penpornk, The fix is modified to make more conservative : [commit](https://github.com/oneapi-src/oneDNN/commit/3a4b5c526805e8c6aaadb1e377b4725988da532b) .  It is in 1.4 ([525 last update](https://github.com/oneapi-src/oneDNN/issues/525#issuecomment-575377402))

**Comment by penpornk** [comment]
@agramesh1 @jojivk73 Thank you for the info!

**Comment by NeoZhangJianyu** [comment]
@npuichigo 
Could you check the answer?
Is it possible to close this issue if it's OK to you?

Thank you!

**Comment by npuichigo** [comment]
@agramesh1 @jojivk73 Thanks for the info. @NeoZhangJianyu It works for me now. I'll close the issue.

**Comment by zhangqiang-hf** [comment]
Dear Sir，
  I use tensorflow2.4.0 with use onednn1.4， test resnet50 infenrence，still slower than not use onednn。
and the onednn  code is：
 https://github.com/oneapi-src/oneDNN/commit/3a4b5c526805e8c6aaadb1e377b4725988da532b

**Comment by Zantares** [comment]
> Dear Sir，
> I use tensorflow2.4.0 with use onednn1.4， test resnet50 infenrence，still slower than not use onednn。
> and the onednn code is：
> [oneapi-src/oneDNN@3a4b5c5](https://github.com/oneapi-src/oneDNN/commit/3a4b5c526805e8c6aaadb1e377b4725988da532b)

The possible reason is running without configuration. Have you ever set below configuration? It's very important to TF with oneDNN:
Env var | value
-- | --
KMP_BLOCKTIME | 0 or 1
KMP_AFFINITY | granularity=fine,verbose,compact,1,0
OMP_NUM_THREADS | Cores/inter
TF_NUM_INTRAOP_THREADS | Cores
TF_NUM_INTEROP_THREADS | 1 or 2

You can get more detail here: https://software.intel.com/content/www/us/en/develop/articles/maximize-tensorflow-performance-on-cpu-considerations-and-recommendations-for-inference.html

**Comment by zhangqiang-hf** [comment]
still slow。
--num_intra_threads=4 --num_inter_threads=2 data_format=NCHW --kmp_affinity=granularity=fine,compact,1,0 --kmp_blocktime=0
or export above setting。

**Comment by wei-v-wang** [comment]
You can try the following on top of what you did previously regarding OpenMP tunings. 

**export TF_ENABLE_MKL_NATIVE_FORMAT=1** (this is latest Intel TF optimization feature and is likely to help with your keras based model). 
python ... as usual.

**Comment by NeoZhangJianyu** [comment]
@zhangqiang-hf 
What's the format of resnet50 model? Keras HD5 or Tensorflow PB?
 
Is it possible to share your test code for this issue? And what's the CPU & OS?

So, we could check it.

**Comment by zhangqiang-hf** [comment]
Dear sir， 
the format of resnet50 is keras HD5。ok， I will share my test code for this issue。
Tensorflow-2.4.0
oneDNN-1.5.0

CPU：mips64
OS：loongnix（fedora21）
python：3.7.9
the test code is as bellow：
[
[resnet50.zip](https://github.com/tensorflow/tensorflow/files/5600036/resnet50.zip)

](url)

Tensorflow build command：
**use mkl**
bazel build --verbose_failures --config=noaws --config=mkl --config=opt --host_copt=-march=loongson3a --action_env=BAZEL_LINKLIBS=-l%:libstdc++.a --cxxopt="-D_GLIBCXX_USE_CXX11_ABI=0" //tensorflow/tools/pip_package:build_pip_package

**not use mkl：**
bazel build --verbose_failures --config=noaws --config=opt --host_copt=-march=loongson3a --action_env=BAZEL_LINKLIBS=-l%:libstdc++.a --cxxopt="-D_GLIBCXX_USE_CXX11_ABI=0" //tensorflow/tools/pip_package:build_pip_package

then ， I test ：python-3.7.9 resnet50.py
the result is:
**resnet50 inference：tensorflow-2.4.0 that use mkl is slower than tensorflow-2.4.0 which not use mkl！**

than, I do the same on x86 platform, it is the same result.

**Comment by NeoZhangJianyu** [comment]
@zhangqiang-hf 
I see your CPU is MIPS64.
Tensorflow with MKL/oneDNN is optimized based on Intel CPU.
We don't know if the optimization is working on MIPS64. 
It's out of the scope of Tensorflow with MKL/oneDNN.

In general, the AVX2, AVX512 intrinsic included in Intel CPU. They are used by MKL/oneDNN to speed up the TF.
If the CPU don't include them, the optimization can't work.

Thank you!

**Comment by zhangqiang-hf** [comment]
> with MKL/oneDNN

Dear sir，
the oneDNN can used on mips64， we have done the job。
Now，
**resnet50 inference：tensorflow-2.4.0 that use mkl is slower than tensorflow-2.4.0 which not use mkl！，also， it is the same phenomenon on x86 i7 platform。x86 still slower when use mkl**
**but， vgg16 and vgg19，tensorflow-2.4.0 that use mkl is quicker than tensorflow-2.4.0 which not use mkl！**

**Comment by NeoZhangJianyu** [comment]
@zhangqiang-hf 
Sorry, I missed to understand your words.

OK, we will check it and feedback you as soon.

Thank you!

**Comment by zhangqiang-hf** [comment]
> @zhangqiang-hf
> Sorry, I missed to understand your words.
> 
> OK, we will check it and feedback you as soon.
> 
> Thank you!

ok，you can use my test case：resnet50.zip to check this issue on x86 Platform :  tensorflow-2..4.0(with mkl or not with mkl) inference,compare the inference performance。
Thank you sir！

**Comment by zhangqiang-hf** [comment]
> @zhangqiang-hf
> Sorry, I missed to understand your words.
> 
> OK, we will check it and feedback you as soon.
> 
> Thank you!

Dear sir，
  Could you reproduce this issue on x86 platform when use the test case？
I‘m eager to check this issue， please give me some oppinion， thank you sir！

**Comment by NeoZhangJianyu** [comment]
@zhangqiang-hf 
Yes, it's ongoing.

Thank you!

**Comment by zhangqiang-hf** [comment]
> @zhangqiang-hf
> Yes, it's ongoing.
> 
> Thank you!

Dear sir，
  How about this issue？Have you reproduced this issue？

**Comment by NeoZhangJianyu** [comment]
@zhangqiang-hf 
I use your parameters to build TF for Intel CPU, there is compile error.
So, I use the default compile parameters. It's OK for non-MKL.

I'm building the 2.4 with MKL now.

I will update if it's finished.

**Comment by NeoZhangJianyu** [comment]
@zhangqiang-hf 
I share you part of progress and guide for you firstly.

After apply the optimization setting as above suggestion, I got some result:
(I got the result by your script. It's not good to test the performance. Like miss the warmup, only run 1 times per case.
So, following result maybe not stable in different test env.)

When BS=1 (your case), the TF with MKL is a litter slower than stack TF.
When BS=10, the performance is same.
When BS=100, 1000, the TF with MKL is better than stack TF obviously.

PS: we think when BS=1, the performance should be better too. The checking is on going.

I test it with Xeon, instead of Core i7. But you can refer to the optimization setting.

```
export TF_ENABLE_MKL_NATIVE_FORMAT=1  
export TF_NUM_INTEROP_THREADS=1
export TF_NUM_INTRAOP_THREADS=24
export OMP_NUM_THREADS=24
export KMP_BLOCKTIME=1
export KMP_AFFINITY=granularity=fine,compact,1,0
```
`TF_ENABLE_MKL_NATIVE_FORMAT` is key optimization, friendly to Keras model inference.
`TF_NUM_INTRAOP_THREADS` & `OMP_NUM_THREADS` are set as the number of cores in the CPU. You could change it for better performance.
`KMP_BLOCKTIME` are set 0, 1 or other number. depend on test result.

When we optimize the Tensorflow inference, the Max Throughput & Latency should be considered in same time.
In this case, the latency of BS=1 and BS = 10 are very close. 
BS =1 and BS=100 are close.
So, you could increase the BS to reach good balance of performance tuning.

**Comment by zhangqiang-hf** [comment]
Thank you sir！
you means BS=100：predict 100 pictures? use for statement excute 100 times? mine is:
def predict_file_path(model, file_path):
    files = os.listdir(file_path)
    index = 0
    start = time.time()
    for f in files:
        try:
            predict_result(model, file_path+'/'+f)
        except:
            print(f)
        index += 1
        if index %100 == 0:
            end = time.time()
            print(index, '-> CPU执行时间(s):', end - start)

could you please share your test code for this case?

**Comment by NeoZhangJianyu** [comment]
@zhangqiang-hf 
Following is the code to create 100 images for BS=100, based on your script.
BS=100 means input 100 images per time, instead of 1 image with 100 times.
```
num_repeats = 100
x = np.tile(x,(num_repeats, 1,1,1))
print('Input image shape:', x.shape)
im = Image.open(file_path)
#im.show()

```

**Comment by NeoZhangJianyu** [comment]
@zhangqiang-hf 

I think your issue is new one for Resnet50 model.
This issue topic is LSTM.

Is it possible to create a new github issue for your model?
You could @me in the new issue, so I can continue following it.

Thank you!

**Comment by zhangqiang-hf** [comment]
> @zhangqiang-hf
> 
> I think your issue is new one for Resnet50 model.
> This issue topic is LSTM.
> 
> Is it possible to create a new github issue for your model?
> You could @me in the new issue, so I can continue following it.
> 
> Thank you!

ok！I will assign a new bug for you！when one pic, why it is slower when use onednn, this  still a issue? am I right?
also, I have checked when bs=100, it is quicker when onednn. but this 100 pictures is the same, when the100 pictures are different, how about?

**Comment by NeoZhangJianyu** [comment]
@zhangqiang-hf 

It's great!
Could you copy your questions in the new issue? I will answer by it.

Thank you!

**Comment by zhangqiang-hf** [comment]
> @zhangqiang-hf
> 
> It's great!
> Could you copy your questions in the new issue? I will answer by it.
> 
> Thank you!
Dear sir,
 I have assign a new issue:
https://github.com/tensorflow/tensorflow/issues/45289

could please continue it?

**Comment by NeoZhangJianyu** [comment]
@zhangqiang-hf 

Yes.

**Comment by Zantares** [comment]
> > @zhangqiang-hf
> > I think your issue is new one for Resnet50 model.
> > This issue topic is LSTM.
> > Is it possible to create a new github issue for your model?
> > You could @me in the new issue, so I can continue following it.
> > Thank you!
> 
> ok！I will assign a new bug for you！when one pic, why it is slower when use onednn, this still a issue? am I right?
> also, I have checked when bs=100, it is quicker when onednn. but this 100 pictures is the same, when the100 pictures are different, how about?

Bad performance is expected if you only run session once because TF with oneDNN will take more time in warmup (1st iteration). You will get similar performance with different pics as long as you don't change pic shape. In same word, TF with oneDNN doesn't care about data, it only cares shape.

**Comment by NeoZhangJianyu** [comment]
@zhangqiang-hf @Zantares 
Yes. 
I just test to run 100 times for BS=1.
The performance of TF with MKL is better than stack TF.

```
    times = 100
    a = time.time()    
    for i in range(times):    
        preds = model.predict(x)
    b = time.time()
    seconds = (b - a)/times
    print("predict() cost time:",seconds)

```
